main/src/ca/uwaterloo/flix/MutationTester.scala

```diff
@@ -0,0 +1,398 @@
+        /**
+        val result = root1.map(r => flix.codeGen(r).unsafeGet)
+        val tests = result.map(res => res.getTests)
+        */
+    }
+
+    private def runMutations(flix: Flix, root: TypedAst.Root, mutatedDefs: List[(Symbol.DefnSym, List[TypedAst.Def])]): Unit = {
+        val totalStartTime = System.nanoTime()
+        var timeTemp = 0.0
+        var survivorCount = 0
+        var mutantCounter = 0
+        mutatedDefs.foreach(mut => {
+            val defName = mut._1.toString
+            val mutationAmount = mut._2.length
+            println(s"testing $defName with $mutationAmount mutations")
+            val defs = root.defs
+            mut._2.foreach(mDef => {
+                mutantCounter += 1
+                val start = System.nanoTime()
+                val n = defs + (mut._1 -> mDef)
+                println(s"mutation: $mDef")
+                val newRoot = root.copy(defs = n)
+                val cRes = flix.codeGen(newRoot).unsafeGet
+                val testResults = cRes.getTests.values.forall(c =>
+                    try {
+                        c.run() match {
+                            case java.lang.Boolean.TRUE => true
+                            case _ => false
+                        }
+                    } catch {
+                        case _: Throwable => false
+                    }
+                )
+                timeTemp += (System.nanoTime() - start).toFloat / 1_000_000_000
+                if (testResults) {
+                    survivorCount += 1
+                    val sym = mDef.sym.toString
+                    println(s"mutation in $sym survived")
+                }
+            })
+        })
+        val totalEndTime = System.nanoTime() - totalStartTime
+        println(s"there where $survivorCount surviving mutations, out of $mutantCounter mutations")
+        val average = timeTemp/mutantCounter
+        println(s"average time to test a mutant:  $average sec")
+        val time = totalEndTime.toFloat/1_000_000_000
+        println(s"total time to test all mutants: $time sec")
+    }
+    private def mutateRoot(root: TypedAst.Root, testee: String): List[(Symbol.DefnSym, List[TypedAst.Def])] = {
+        val defs = root.defs
+        val defSyms = root.modules.filter(pair => (pair._1.toString.equals(testee))).values.toList.flatten
+        mutateDefs(defs, defSyms).flatten
+    }
+
+
+    private def mutateDefs(defs: Map[Symbol.DefnSym, TypedAst.Def], defSyms: List[Symbol]) = {
+        defs.toList.map(d => (d._1, d._2) match {
+            case (s, fun) =>
+                if (defSyms.contains(s)) {
+                    val mutExps = mutateExpr(fun.exp)
+                    val mutDefs = mutExps.map(mexp => fun.copy(exp = mexp))
+                    Some(d._1 -> mutDefs)
+                } else None
+            case _ => None
+        })
+    }
+
+
+    private def mutationPermutations[A](exps: List[A], mutationFunc: (A => List[A])): List[List[A]] = {
```

Comment:
Why not just have a function that returns all permutations? Then you can map over it. You don't have to have such a complicated function as this.